### PR TITLE
Permit partial iteration in On Demand

### PR DESCRIPTION
**Pull request by jkeiser**
## Overview

simdjson's On Demand functionality requires each array and object to be fully consumed (iterated) before the next value in the file can be used. This is not a problem when the user actually iterates over everything they use. However, if the user terminates iteration of an array or object early (for example, when retrieving just one object field), simdjson has to detect this, and skip to the end of the value, completing iteration.

Currently this process is eager, handled in the destructor. This causes some simple things not to work, leads to bad error messages, and forces the user to understand and reason about destruction order and move semantics.

This PR switches partial iteration to be more lazy, only finishing up iteration when the user *asks* for another value, which solves the problems.

Performance looks like roughly a wash: there are some weird bumps both directions that go away when you run it over and over, but generally you can see the picture from the numbers below.

## Problem

Right now, simdjson eagerly detects and completes iteration of each object and array in its destructor. This works fine, but has a few issues:

* Programs that terminate early (for instance, looking at a single value or the first N records in a JSON file) waste time looping over the rest of the JSON.
* Some straightforward constructs--such as `object["a"]["b"]`--do not work because of temporary object destruction order.
* Users need to ensure proper destructor order by tweaking their variable nesting and declaration order. This means one extra barrier to pass before one can use simdjson.
* This forces move semantics on everything in On Demand, causing yet *another* barrier for users to cross.
* Move semantics yield pretty bad error messages: it's hard to tell just *why* `object["a"]["b"]` is failing.

## Solution

With this PR, we don't finish iterating children until the user actually asks for another value in a parent array or object.

Because the user could stop anywhere (even several levels deep), we track the depth of the iterator in the JSON at all times. When the user asks for the next value in an outer array/object, it checks whether it's already at the parent depth, meaning all children have been fully iterated. If not, it skips JSON tokens until it ascends to the parent depth.

By removing the destructors, we also remove move semantics. Not only does this solve most of the usability issues, it is also theoretically more performant, because it can use default copy and assignment constructors, which supposedly are more optimizeable by the compiler. (This remains to be seen, but results so far are good!)

The con side of this is, programs are a little less safe: nothing prevents you from keeping a value around between two iterations of an array/object, which will yield unexpected results if you try to use it. There may well be ways to add safety rails for this, but it may be worth the pain in the meantime to alleviate some of the usability issues caused by the current regime.

### skylake-x clang 10.0.1

```
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Branch  Benchmark                         Time             CPU   Iterations best_branch_miss best_bytes_per_sec best_cache_miss best_cache_ref best_cycles best_cycles_per_byte best_docs_per_sec best_frequency best_instructions best_instructions_per_byte best_instructions_per_cycle best_items_per_sec branch_miss      bytes bytes_per_second cache_miss  cache_ref     cycles cycles_per_byte docs_per_sec  frequency instructions instructions_per_byte instructions_per_cycle      items items_per_second
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
PR      PartialTweets<OnDemand>      203824 ns       203823 ns         3425           1.658k           3.36882G               2         5.503k    598.761k             0.948134           5.3345k       3.19409G          1.80783M                    2.86269                     3.01929            533.45k    1.79348k   631.515k       2.88556G/s    1.88964   5.50799k     603.5k        0.955638   4.90622k/s  2.9609G/s     1.80783M               2.86269                2.99558        100       490.622k/s [best: throughput=  3.37 GB/s doc_throughput=  5334 docs/s instructions=     1807832 cycles=      598761 branch_miss=    1658 cache_miss=       2 cache_ref=      5503 items=       100 avg_time=    189076 ns]
master  PartialTweets<OnDemand>      206106 ns       206103 ns         3393           1.915k           3.35229G               0         6.525k    601.746k             0.952861          5.30833k       3.19427G          1.90674M                    3.01931                     3.16868           530.833k    2.09923k   631.515k       2.85365G/s   0.666667   6.54689k   609.809k        0.965629   4.85195k/s 2.95877G/s     1.90674M               3.01931                3.12678        100       485.195k/s [best: throughput=  3.35 GB/s doc_throughput=  5308 docs/s instructions=     1906742 cycles=      601746 branch_miss=    1915 cache_miss=       0 cache_ref=      6525 items=       100 avg_time=    191082 ns]
PR      LargeRandom<OnDemand>      70093969 ns     70091374 ns           10         967.269k           657.148M         3.4465M       3.57309M    223.243M              4.85324           14.2862        3.1893G          619.118M                    13.4595                     2.77329           14.2862M    968.075k   45.9988M       625.867M/s   3.48876M   3.57315M   223.446M         4.85765    14.2671/s 3.18792G/s     619.118M               13.4595                2.77077      1000k       14.2671M/s [best: throughput=  0.66 GB/s doc_throughput=    14 docs/s instructions=   619118370 cycles=   223243220 branch_miss=  967269 cache_miss= 3446496 cache_ref=   3573088 items=   1000000 avg_time=  70069116 ns]
master  LargeRandom<OnDemand>      69909773 ns     69902107 ns           10         948.979k           658.679M        3.46399M       3.57396M    222.649M              4.84033           14.3195       3.18822G          630.119M                    13.6986                      2.8301           14.3195M    950.621k   45.9988M       627.561M/s   3.48671M   3.57369M    222.82M         4.84405    14.3057/s 3.18761G/s     630.119M               13.6986                2.82793      1000k       14.3057M/s [best: throughput=  0.66 GB/s doc_throughput=    14 docs/s instructions=   630119355 cycles=   222649077 branch_miss=  948979 cache_miss= 3463987 cache_ref=   3573957 items=   1000000 avg_time=  69883466 ns]
PR      Kostya<OnDemand>           67279343 ns     67272543 ns           10         457.288k            2.0433G         6.4558M       6.47123M    213.875M              1.55766           14.8815       3.18278G          607.522M                    4.42461                     2.84054           7.80218M    457.183k   137.305M       1.90085G/s   6.35829M   6.47068M   214.412M         1.56158    14.8649/s 3.18722G/s     607.522M               4.42461                2.83343   524.288k       7.79349M/s [best: throughput=  2.04 GB/s doc_throughput=    14 docs/s instructions=   607521614 cycles=   213875075 branch_miss=  457288 cache_miss= 6455799 cache_ref=   6471230 items=    524288 avg_time=  67252572 ns]
master  Kostya<OnDemand>           67315426 ns     67312753 ns           10         455.869k            2.0432G        6.31429M       6.47058M    214.305M               1.5608           14.8807       3.18902G          630.066M                     4.5888                     2.94004           7.80178M    456.116k   137.305M       1.89972G/s    6.3568M   6.47057M   214.541M         1.56251     14.856/s 3.18723G/s     630.066M                4.5888                2.93681   524.288k       7.78884M/s [best: throughput=  2.04 GB/s doc_throughput=    14 docs/s instructions=   630065819 cycles=   214305300 branch_miss=  455869 cache_miss= 6314286 cache_ref=   6470577 items=    524288 avg_time=  67288524 ns]
PR      DistinctUserID<OnDemand>     210719 ns       210718 ns         3325           2.318k           3.26545G               7         2.979k    617.656k             0.978054          5.17082k       3.19379G          1.77011M                    2.80296                     2.86585           594.644k    2.51158k   631.515k       2.79115G/s    7.52782   2.99231k   623.703k        0.987629   4.74568k/s 2.95989G/s     1.77011M               2.80296                2.83807        115       545.753k/s [best: throughput=  3.27 GB/s doc_throughput=  5170 docs/s instructions=     1770112 cycles=      617656 branch_miss=    2318 cache_miss=       7 cache_ref=      2979 items=       115 avg_time=    195425 ns]
master  DistinctUserID<OnDemand>     206202 ns       206200 ns         3391           1.566k           3.32969G               0         1.416k    605.747k             0.959197          5.27254k       3.19382G           1.8642M                    2.95195                     3.07752           606.342k    1.67402k   631.515k        2.8523G/s  0.0312592   1.54589k   609.649k        0.965376   4.84965k/s 2.95659G/s      1.8642M               2.95195                3.05782        115        557.71k/s [best: throughput=  3.33 GB/s doc_throughput=  5272 docs/s instructions=     1864198 cycles=      605747 branch_miss=    1566 cache_miss=       0 cache_ref=      1416 items=       115 avg_time=    191018 ns]
PR      FindTweet<OnDemand>          147320 ns       147319 ns         4747              880            4.8055G               0         1.959k    419.787k              0.66473          7.60948k       3.19436G          1.35302M                     2.1425                     3.22311           7.60948k     936.443   631.515k       3.99231G/s   1.89593m   2.04981k   422.129k        0.668438   6.78798k/s  2.8654G/s     1.35302M                2.1425                3.20523          1       6.78798k/s [best: throughput=  4.81 GB/s doc_throughput=  7609 docs/s instructions=     1353020 cycles=      419787 branch_miss=     880 cache_miss=       0 cache_ref=      1959 items=         1 avg_time=    132236 ns]
```

### skylake-x gcc 10.2.0

```
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Benchmark                         Time             CPU   Iterations best_branch_miss best_bytes_per_sec best_cache_miss best_cache_ref best_cycles best_cycles_per_byte best_docs_per_sec best_frequency best_instructions best_instructions_per_byte best_instructions_per_cycle best_items_per_sec branch_miss      bytes bytes_per_second cache_miss  cache_ref     cycles cycles_per_byte docs_per_sec  frequency instructions instructions_per_byte instructions_per_cycle      items items_per_second
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
PR      PartialTweets<OnDemand>      242552 ns       242547 ns         2886           2.271k           2.79494G               0         4.193k    721.577k              1.14261          4.42578k       3.19354G          2.21965M                     3.5148                     3.07611           442.578k    2.41176k   631.515k       2.42486G/s  0.0824671    4.2497k   726.807k         1.15089   4.12291k/s 2.99656G/s     2.21965M                3.5148                3.05398        100       412.291k/s [best: throughput=  2.79 GB/s doc_throughput=  4425 docs/s instructions=     2219652 cycles=      721577 branch_miss=    2271 cache_miss=       0 cache_ref=      4193 items=       100 avg_time=    227741 ns]
master  PartialTweets<OnDemand>      240529 ns       240524 ns         2908           2.216k           2.82197G               1         5.502k    714.731k              1.13177          4.46857k       3.19383G          2.20691M                    3.49462                     3.08774           446.857k    2.34097k   631.515k       2.44526G/s   0.674347   5.50157k   720.071k         1.14023   4.15759k/s 2.99376G/s     2.20691M               3.49462                3.06485        100       415.759k/s [best: throughput=  2.82 GB/s doc_throughput=  4468 docs/s instructions=     2206905 cycles=      714731 branch_miss=    2216 cache_miss=       1 cache_ref=      5502 items=       100 avg_time=    225635 ns]
PR      LargeRandom<OnDemand>      70542150 ns     70540429 ns           10         891.182k           652.474M        3.43501M       3.57268M     224.84M              4.88796           14.1846       3.18927G          640.959M                    13.9343                     2.85073           14.1846M    891.248k   45.9988M       621.882M/s   3.47993M   3.57278M   224.862M         4.88844    14.1763/s 3.18771G/s     640.959M               13.9343                2.85045      1000k       14.1763M/s [best: throughput=  0.65 GB/s doc_throughput=    14 docs/s instructions=   640958910 cycles=   224840341 branch_miss=  891182 cache_miss= 3435005 cache_ref=   3572680 items=   1000000 avg_time=  70516585 ns]
master  LargeRandom<OnDemand>      65519449 ns     65517799 ns           11         868.521k           703.532M        3.48205M        3.5723M    208.504M              4.53282           15.2946       3.18898G          641.959M                     13.956                     3.07888           15.2946M     868.21k   45.9988M       669.556M/s   3.47139M   3.57237M   208.849M         4.54031     15.263/s 3.18766G/s     641.959M                13.956                 3.0738      1000k        15.263M/s [best: throughput=  0.70 GB/s doc_throughput=    15 docs/s instructions=   641958637 cycles=   208503947 branch_miss=  868521 cache_miss= 3482050 cache_ref=   3572302 items=   1000000 avg_time=  65494531 ns]
PR      Kostya<OnDemand>           71284303 ns     71282250 ns           10         471.585k           1.93237G        6.39509M       6.47132M    226.569M              1.65011           14.0735       3.18862G          685.012M                    4.98898                     3.02342           7.37859M    472.291k   137.305M       1.79393G/s   6.37209M   6.47135M   227.208M         1.65477    14.0287/s 3.18744G/s     685.012M               4.98898                3.01491   524.288k        7.3551M/s [best: throughput=  1.93 GB/s doc_throughput=    14 docs/s instructions=   685012178 cycles=   226568778 branch_miss=  471585 cache_miss= 6395093 cache_ref=   6471316 items=    524288 avg_time=  71257865 ns]
master  Kostya<OnDemand>           72312612 ns     72310372 ns           10         463.765k           1.90228G        6.36867M       6.47043M     230.18M              1.67642           13.8544       3.18902G          679.769M                     4.9508                      2.9532           7.26372M    464.097k   137.305M       1.76842G/s   6.37662M   6.47043M     230.5M         1.67874    13.8293/s 3.18764G/s     679.769M                4.9508                2.94911   524.288k       7.25052M/s [best: throughput=  1.90 GB/s doc_throughput=    13 docs/s instructions=   679769177 cycles=   230180448 branch_miss=  463765 cache_miss= 6368668 cache_ref=   6470433 items=    524288 avg_time=  72286879 ns]
PR      DistinctUserID<OnDemand>     222199 ns       222195 ns         3153           1.992k           3.06914G               0         3.564k    657.158k              1.04061          4.85996k       3.19376G          2.17925M                    3.45083                     3.31618           558.895k    2.12423k   631.515k       2.64697G/s   0.019981   3.61878k   661.129k         1.04689   4.50055k/s 2.97544G/s     2.17925M               3.45083                3.29626        115       517.563k/s [best: throughput=  3.07 GB/s doc_throughput=  4859 docs/s instructions=     2179253 cycles=      657158 branch_miss=    1992 cache_miss=       0 cache_ref=      3564 items=       115 avg_time=    207163 ns]
master  DistinctUserID<OnDemand>     232220 ns       232214 ns         3013           2.288k           2.93747G               0         1.571k    686.552k              1.08715          4.65147k       3.19347G          2.16095M                    3.42185                     3.14754           534.919k    2.43638k   631.515k       2.53276G/s  0.0212413   1.52857k   691.597k         1.09514   4.30637k/s 2.97827G/s     2.16095M               3.42185                3.12458        115       495.232k/s [best: throughput=  2.94 GB/s doc_throughput=  4651 docs/s instructions=     2160951 cycles=      686552 branch_miss=    2288 cache_miss=       0 cache_ref=      1571 items=       115 avg_time=    216704 ns]
PR      FindTweet<OnDemand>          163648 ns       163634 ns         4270           1.218k           4.28035G               0         1.844k    471.301k             0.746302          6.77791k       3.19444G          1.57058M                    2.48701                     3.33244           6.77791k    1.30259k   631.515k       3.59427G/s   0.012178   1.90099k   474.197k        0.750888    6.1112k/s 2.89792G/s     1.57058M               2.48701                3.31209          1        6.1112k/s [best: throughput=  4.28 GB/s doc_throughput=  6777 docs/s instructions=     1570582 cycles=      471301 branch_miss=    1218 cache_miss=       0 cache_ref=      1844 items=         1 avg_time=    148583 ns]
```

### CODE REVIEW

**Comment by lemire** [comment]
Sounds really good. Marking for 0.8.0.

**Comment by jkeiser** [comment]
Note: while the performance boost appears to be real, the *size* of the performance boost depends significantly on the size/layout of the benchmark executable. Adding unrelated benchmarks caused the different to shrink from 6% to 2% (and the final version to run at 3.4GB/s instead of 3.6).

**Comment by jkeiser** [comment]
Regardless, I've added the tests that needed to be added (and found a bug!), and made `object["a"]["b"]` work. Working through multiplatform issues now, then it'll be ready for review.

**Comment by jkeiser** [comment]
@lemire note document_stream_tests failed with some kind of memory dump--not that it's impossible my changes caused an issue, but they are pretty far away from that code in general ... https://cloud.drone.io/simdjson/simdjson/7343/10/2

**Comment by jkeiser** [comment]
All tests are passing, and we are still a bit faster (it was even faster in earlier tests, and I think is strongly affected by binary layout).  It has solid tests, and updated documentation. I'll post more numbers soon, but this is ready for review.

**Comment by lemire** [comment]
@jkeiser Trailing white spaces!!!

**Comment by jkeiser** [comment]
@lemire will fix :) Sounds like I must have introduced some in the doc update.

File: benchmark/bench_ondemand.cpp [hunk 1]
```diff
@@ -8,19 +8,22 @@ SIMDJSON_PUSH_DISABLE_ALL_WARNINGS
 SIMDJSON_POP_DISABLE_WARNINGS
 
 #include "partial_tweets/ondemand.h"
-#include "partial_tweets/iter.h"
+// #include "partial_tweets/iter.h"
```
Comment (lemire) [inline benchmark/bench_ondemand.cpp:11]: We probably want to clean out dead comments in time.

File: benchmark/distinctuserid/ondemand.h [hunk 2]
```diff
@@ -35,25 +35,13 @@ simdjson_really_inline bool OnDemand::Run(const padded_string &json) {
...
+      ids.push_back(tweet["user"]["id"]);
       // Not all tweets have a "retweeted_status", but when they do
       // we want to go and find the user within.
       auto retweet = tweet["retweeted_status"];
       if(!retweet.error()) {
-          ondemand::object retweet_content = retweet;
-          ondemand::object reuser = retweet_content["user"];
-          int64_t rid = reuser["id"];
-          ids.push_back(rid);
+          ids.push_back(retweet["user"]["id"]);
```
Comment (lemire) [inline benchmark/distinctuserid/ondemand.h:44]: So.... soo.... soo... much prettier.

File: benchmark/bench_ondemand.cpp (hunk 1 above)
Comment (jkeiser) [reply benchmark/bench_ondemand.cpp:11]: Oh yeah! I wanted them there to make it clear I wasn't maintaining the iterator versions, but I'll remove these comments before we merge.

I changed how iteration worked; I actually think it'll be worth restoring raw iteration, but my priority here was a safe `value_iterator` that tracked its own depth.

**Comment by jkeiser** [comment]
@lemire your comment about the pretty reminded me of a few other small cleanups I wanted to make after this change. [This commit(https://github.com/simdjson/simdjson/pull/1330/commits/806cb39103c0b0d2f1e9d77fc5914f88131ff425) shows them (nothing big, just quality of life).

File: benchmark/distinctuserid/ondemand.h (hunk 2 above)
Comment (jkeiser) [reply benchmark/distinctuserid/ondemand.h:44]: 🥳

**Comment by lemire** [comment]
@jkeiser This PR looks good. It is always best for me to actually work with the code to judge it properly, but I think you can merge it at will.

**Comment by jkeiser** [comment]
Ok! I am working through the next PR in my head: I want to restore the ability to validate that you are using it right.

**Pull request merged**